src/game/common/rts/team.cpp

```diff
@@ -0,0 +1,815 @@
+        Utf8String str;
+        str.Format("%s%d", g_theNameKeyGenerator->Key_To_Name(g_teamGenericScriptHookKey).Str(), key);
+        m_genericScripts[key] = d->Get_AsciiString(Name_To_Key(str), &exists);
+
+        if (!exists) {
+            m_genericScripts[key].Clear();
+        }
+    }
+
+    m_transport = d->Get_AsciiString(g_teamTransportKey, &exists);
+    m_transportsExit = d->Get_Bool(g_teamTransportsExitKey, &exists);
+    m_teamStartsFull = d->Get_Bool(g_teamStartsFullKey, &exists);
+    m_reifnrocementOrigin = d->Get_AsciiString(g_teamReinforcementOriginKey, &exists);
+    m_veterancy = d->Get_Int(g_teamVeterancyKey, &exists);
+}
+
+void TeamTemplateInfo::Xfer_Snapshot(Xfer *xfer)
+{
+    uint8_t version = 1;
+    xfer->xferVersion(&version, 1);
+    xfer->xferInt(&m_productionPriority);
+}
+
+TeamPrototype::TeamPrototype(
+    TeamFactory *tf, const Utf8String &name, Player *owner_player, bool is_singleton, Dict *d, unsigned int id) :
+    m_factory(tf),
+    m_owningPlayer(owner_player),
+    m_ID(id),
+    m_name(name),
+    m_productionConditionAlwaysFalse(false),
+    m_productionConditionScript(nullptr),
+    m_teamTemplate(d)
+{
+    captainslog_dbgassert(m_owningPlayer != nullptr, "bad args to TeamPrototype ctor");
+
+    if (m_factory != nullptr) {
+        m_factory->Add_Team_Prototype_To_List(this);
+    }
+
+    if (m_owningPlayer != nullptr) {
+        m_owningPlayer->Add_Team_To_List(this);
+    }
+
+    m_genericScriptsInitalized = false;
+
+    for (int i = 0; i < ARRAY_SIZE(m_genericScripts); i++) {
+        m_genericScripts[i] = nullptr;
+    }
+}
+
+void Delete_Team_Callback(Team *o)
+{
+    if (o != nullptr) {
+        g_theTeamFactory->Team_About_To_Be_Deleted(o);
+        o->Delete_Instance();
+    }
+}
+
+TeamPrototype::~TeamPrototype()
+{
+    Remove_All_Team_Instance_List(Delete_Team_Callback);
+
+    if (m_owningPlayer != nullptr) {
+        m_owningPlayer->Remove_Team_From_List(this);
+    }
+
+    if (m_productionConditionScript != nullptr) {
+        m_productionConditionScript->Delete_Instance();
+    }
+
+    m_productionConditionScript = nullptr;
+
+    for (int i = 0; i < ARRAY_SIZE(m_genericScripts); i++) {
+        if (m_genericScripts[i] != nullptr) {
+            m_genericScripts[i]->Delete_Instance();
+            m_genericScripts[i] = nullptr;
+        }
+    }
+}
+
+void TeamPrototype::Xfer_Snapshot(Xfer *xfer)
+{
+    unsigned char version = 2;
+    xfer->xferVersion(&version, 2);
+
+    int index;
+
+    if (xfer->Get_Mode() == XFER_SAVE) {
+        index = m_owningPlayer->Get_Player_Index();
+    }
+
+    xfer->xferInt(&index);
+
+    m_owningPlayer = g_thePlayerList->Get_Nth_Player(index);
+
+    if (version >= 2) {
+        xfer->xferAsciiString(&m_attackPriorityName);
+    }
+
+    xfer->xferBool(&m_productionConditionAlwaysFalse);
+    xfer->xferSnapshot(&m_teamTemplate);
+    unsigned short size = 0;
+    DLINK_ITERATOR<Team> iter = Iterate_Team_Instance_List();
```

Comment:
If this `iter` is exclusive to the loop below, I suggest put it in scope to avoid reuse of this iterator:

```c++
{
    DLINK_ITERATOR<Team> iter = Iterate_Team_Instance_List();
    while (!iter.Done()) {
        size++;
        iter.Advance();
    }
}
```